src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,516 @@
+
+	pace_input.pin_id = PACE_PIN_ID_CAN;
+	pace_input.pin = (u8 *)can;
+	pace_input.pin_length = strlen(can);
+
+	/* EF.CardAccess can only be read from MF */
+	sc_format_path("3F00", &path);
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &path, NULL),
+		"sc_select_file failed");
+
+	LOG_TEST_RET(card->ctx, perform_pace(card, pace_input, &pace_output, EAC_TR_VERSION_2_02), 
+		"Error verifying CAN");
+
+	return SC_SUCCESS;
+}
+
+static int eoi_get_data(sc_card_t *card, u8 data_id, u8 *buf, size_t len)
+{
+	int r;
+	sc_apdu_t apdu;
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0xCA, 0x01, data_id);
+	apdu.resp = buf;
+	apdu.resplen = len;
+	apdu.le = len;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	return r;
+}
+
+#define ATR_MATCH 1
+
+static int eoi_match_card(sc_card_t* card) {
+	LOG_FUNC_CALLED(card->ctx);
+	if (_sc_match_atr(card, eoi_atrs, &card->type) >= 0) {
+		sc_log(card->ctx, "ATR recognized as Slovenian eID card");
+		LOG_FUNC_RETURN(card->ctx, ATR_MATCH);
+	}
+	LOG_FUNC_RETURN(card->ctx, !ATR_MATCH);
+}
+
+static int eoi_init(sc_card_t* card) {
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	u8 version[6];
+	size_t i, j;
+	scconf_block **found_blocks, *block;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (eoi_get_data(card, 0x16, version, sizeof(version)) != SC_SUCCESS)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_WRONG_CARD);
+
+	if (privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	privdata = calloc(1, sizeof(struct eoi_privdata));
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+	card->drv_data = privdata;
+
+	sprintf(privdata->version, "%X%02X.%02X%02X", version[0], version[1], version[2], version[3]);
+	sc_log(card->ctx, "App version: %s", privdata->version);
+
+	memset(&card->sm_ctx, 0, sizeof card->sm_ctx);
+	card->sm_ctx.ops.open = eoi_sm_open;
+
+	card->max_send_size = SC_MAX_APDU_RESP_SIZE;
+	card->max_recv_size = SC_MAX_APDU_RESP_SIZE;
+
+	for (i = 0; i < sizeof eoi_curves / sizeof * eoi_curves; ++i) {
+		LOG_TEST_RET(card->ctx, _sc_card_add_ec_alg(
+			card, eoi_curves[i].len,
+			SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ECDSA_HASH_NONE,
+			0, &eoi_curves[i].oid
+		), "Add EC alg failed");
+	}
+
+	privdata->can = getenv("EOI_CAN");
+	for (i = 0; card->ctx->conf_blocks[i]; i++) {
+		found_blocks = scconf_find_blocks(card->ctx->conf, card->ctx->conf_blocks[i],
+					"card_driver", "eoi");
+		if (!found_blocks)
+			continue;
+
+		for (j = 0, block = found_blocks[j]; block; j++, block = found_blocks[j]) {
+			if (!privdata->can)
+				privdata->can = scconf_get_str(block, "can", NULL);
+		}
+	}
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int eoi_finish(sc_card_t* card)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (privdata) {
+		memset(privdata, 0, sizeof(struct eoi_privdata));
```

Comment:
Since you're storing the CAN in the private data, I suggest you use sc_mem_secure_alloc and sc_mem_secure_free for handling the buffer

Reply:
Done.